### DISCUSSION TITLE
V6: Aligning guidelines

### DISCUSSION THREAD

**Discussion by zhe-he** [fabricjs/fabric.js#10033] [Show and tell]
I wrote a v6 version of the reference line plugin based on the original reference line, which supports dragging, zooming, and resizing the display of reference lines.

[Demo Access](https://codesandbox.io/p/devbox/fabric-aligning-guidelines-4pr6x7?file=%2Fsrc%2Ftest.tsx%3A23%2C26)
[Alternative Access](https://stackblitz.com/edit/fabric-lib-aligning-guidelines?file=src%2Ftest.tsx&terminal=dev)

1. Fix the bug where reference lines become unusable when the canvas is zoomed in too much or too little.

https://github.com/user-attachments/assets/fc026f55-7756-4862-b235-ee1ca8493285

2.Fix the issue of 1px line segment deviation.

https://github.com/user-attachments/assets/ae943b85-b17d-481a-8d44-4ab1f8b94168

3. Add new alignments (left-center, left-right, center-left, center-right, right-left, right-center) based on the existing alignments (left-left, center-center, right-right).

https://github.com/user-attachments/assets/05b940dd-07f5-4eb7-8aad-97975dafb099

4. Fix the bug in the old version where shapes cannot be fixed when horizontal and vertical auxiliary lines appear.

https://github.com/user-attachments/assets/f40bf5b7-7818-49c7-83d9-abee4a162451

5. Add reference lines that appear during resizing and scaling.

https://github.com/user-attachments/assets/986e8df0-de5e-4a38-be86-582ceffb8cd1

HOW TO USE: Use the same method as the old version.
```typescript
import { initAligningGuidelines } from "./lib/aligning-guidelines";

initAligningGuidelines(canvas)
```

**Comment by Smrtnyk** [comment]
Looks amazing

**Comment by usamshen** [comment]
@zhe-he would love to use this in boardx, is this under MIT license? thanks for the great work!

**Reply by zhe-he** [reply]
Yes.
You can use it directly. If there are any bugs, please let me know, and I will make timely modifications once I see them.

**Reply by asturur** [reply]
If you want to contribute it you can open a PR @zhe-he i m not sure how this would be packaged yet, that is also why original guidelines are still there untouched

**Comment by zhe-he** [comment]
@asturur 
I have submitted the PR.
#10043

**Comment by usamshen** [comment]
<img width="664" alt="Screenshot 2024-08-16 at 10 12 58 PM" src="https://github.com/user-attachments/assets/f3123d8f-0500-4c3d-add3-226599c6a127">

@zhe-he  

Issue:
When zooming out on the board, multiple lines of guidelines become visible. However, when zooming in, the same guidelines are not displayed as multiple lines.

Expected Behavior:
Guidelines should be consistently visualized regardless of the zoom level. The behavior should follow the same rules whether zooming in or out.

Steps to Reproduce:

1. Zoom out on the board.
2. Observe the visualization of multiple guideline lines.
3. Zoom in to the same context.
4. Notice the discrepancy where multiple lines are no longer visible.
5. Additional Information:

Fabric.js version: V 6

Proposed Solution:
Ensure that guidelines are rendered consistently across all zoom levels, following the same visual rules.

**Reply by asturur** [reply]
You mean you would like to align with objects that are out of view?

**Reply by usamshen** [reply]
The current setting is appropriate. It should only apply to the visible widgets.

The question pertains to the unit of measurement for conflicts or tolerance in visualizing lines or snaps. Should this be based on the original pixel size or the zoomed size?

**Reply by asturur** [reply]
mmm good question.
In my opinion the guildelines and snapping logic requires a lot of options to make sense to everyone.

I think my first idea would be that you see different alignment if they do not fall in the same screen pixel at that moment, not css pixel, not canvas pixel.

The idea would be that if there is no a pixel space to draw it, is the same line

**Reply by zhe-he** [reply]
@usamshen 
he reason for this issue is that off-screen graphics are not drawn and therefore are not included in the default graphics comparison set.
This problem has now been resolved. You need to manually set the comparison set for the graphics. Please download the latest code and use getObjectsByTarget:
```ts
// You can customize the return graphic, and the example will only compare it with sibling elements
new AligningGuidelines(myCanvas, {
    getObjectsByTarget: function (target) {
        const set = new Set<FabricObject>();
        const p = target.parent ?? target.canvas;
        p?.getObjects().forEach((o) => {
            set.add(o);
        });
        // Please remember to exclude yourself, or you will always align with yourself.
        set.remove(target);
        return set;
    },
});
```
